### PR TITLE
Add MaybeMatcher

### PR DESCRIPTION
**Pull request by gabfssilva**
Pretty much the same as the `ResultMatcher` tool, but for the `Maybe` monad:

```ruby
Dry::Matcher::MaybeMatcher.(Maybe(2)) do |m|
  m.some(Integer) do |i|
    "Got int: #{i}"
  end

  m.some do |v|
    "Yay: #{v}"
  end

  m.none do
    "Boo: none"
  end
end # Got int: 2
```

### CODE REVIEW

File: lib/dry/matcher/match.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+require 'dry/matcher'
+
+module Dry
+  class Matcher
+    Match = ::Proc.new do |value, patterns|
```
Comment (timriley) [inline lib/dry/matcher/match.rb:7]: Perhaps this could be called `PatternMatch`, to make its purpose clearer?

**Review by timriley — CHANGES_REQUESTED** [review]
Seems like a good addition to me! @flash-gordon, @solnic, what do you think?

Left you a small naming suggestion :)

File: lib/dry/matcher/match.rb (hunk 1 above)
Comment (gabfssilva) [reply lib/dry/matcher/match.rb:7]: Sure, I'll change it.

Comment (gabfssilva) [reply lib/dry/matcher/match.rb]: Done.

**Comment by flash-gordon** [comment]
You folks do know ruby has pattern matching built-in, right? :)

**Comment by solnic** [comment]
What @flash-gordon said. Personally, I think this gem should be already marked as something that is slowly fading away. I'm OK with adding this feature but it would be good to make it very clear that now people should be switching to native PM. I know it's marked as experimental but in the long term it'll become stable, probably in MRI 3.1.

**Comment by flash-gordon** [comment]
PM is not experimental starting ruby 3.0

**Comment by timriley** [comment]
While dry-matcher still exists, this is a good and helpful addition :) Thanks @gabfssilva!

I'll cut a release in the coming week.

**Comment by flash-gordon** [comment]
Yeah, I'm just pointing out you'll have a better UX with a built-in feature. It's more general and simply shorter (I bet it's faster too).

**Pull request merged**